keps/sig-scheduling/5710-workload-aware-preemption/README.md

```diff
@@ -0,0 +1,1064 @@
+(we can absolutely handle frequent updates to `Workload.Spec.PreemptionPriorityClassName`
+but we can't handle updating potentially hundreds of thousands of pods within that workload
+that frequently). In this case, we limit ourselves to documentation.
+
+```
+<<[UNRESOLVED priority status]>>
+We should introduce/describe `workload.status` to reflect:
+- actual priority (a number) that is inferred from the priority class
+- express the preemption priority that is currently used by scheduler to reflect that
+  it aknowledged the change
+<<[/UNRESOLVED]>>
+```
+
+### Preemption algorithm
+
+We start with describing at the high-level how existing pod-level preemption algorithm works.
+Below, we will show how to generalize it to workloads.
+
+To check if a pod P can be scheduled on a given node with preemption we:
+
+1. Identify the list of potential victims - all running pods with priority lower than the new pod P.
+
+1. If removing all these victims would not make the node feasible, the node is infeasible.
+   
+1. From the list of potential victims, we try to reprieve (remove from the victims list) any pods
+   whose eviction would violate PodDisruptionBudget.
+
+1. From remaining potential victims, we start to reprieve pods starting from the highest priority
+   and working down until the set of remaining victims still keeps the node feasible.
+
+Once we compute the feasibility and list of victims for all nodes, we score that and choose the
+best options.
+
+The above algorithm achieves our principles, as by eliminating highest priority pods first, it
+effectively tries to minimize the cascading preemptions later.
+
+
+We want to generalize the same algorithm to the workload case. However, the difference is not only
+moving to the level of `Workload`, but also no longer operating at the level of individual nodes.
+We need to look at the cluster as a whole. With that in mind, keeping the algorithm efficient
+becomes a challenge, thus we modify to the approach below.
+
+To check if a workload W can be scheduled on a given cluster with preemption we:
+
+1. Identify the list of potential victims:
+   - all running workloads with (preemption) priority lower than the new workload W
+   - all individual pods (not being part of workloads) with priority lower than the new workload W
+
+1. If removing all the potential victims would not make the new workload W schedulable,
+   the workload is unschedulable even with preemption.
+
+```
+<<[UNRESOLVED PodDisruptionBudget violations]>>
+How critical is reprieving workloads and pods violating PodDisruptionBudgets? We no longer can
+afford full workload scheduling trying to reprieve every individual pod and workload.
+
+We could consider finding the first one that can't be reprieved using binary search, but if we
+can't reprieve any of those, learning about that would require O(N) full workload schedulings
+with N being number of workload/pods violating PDB.
+<<[/UNRESOLVED]>>
+```
```

Comment:
Let's assume that nodes are either high-pod-per node count, or low pod-per-node count.  Its a bimodal distribution.

Let's further assume that if Gang scheduling is used, then the node is going to usually be low pod-per-node count.

So, then we can do the following:
1. Individual Pod as preemptor - assume high pod-per-node,  use current algorithm, which is optimized for many pods per node, consider all victims.
1. Gang as preemptor - assume low pod-per-node in all cases, consider a maximum of e.g. 4 reprieves per node, to keep compute time down, and just stop reprieving in the case where there are more things on the node.,

Reply:
LGTM